### DISCUSSION TITLE
python多執行緒的簡單應用

### DISCUSSION THREAD

**Discussion by sean1325418** [1101-GS4719-C/Python-Community-GS4719-C#396] [General]
因為最近有在研究怎麼讓python同時做兩件事也就是多執行緒，剛好上課也有提到就來分享一下最簡單的應用(其實是也不會難的)。
```
import threading
import time

def print_hello():  #副函式
    for i in range(3):  #做3次
        time.sleep(0.5)  #暫停0.5秒
        print("Hello " + str(i))

def print_hellowrold(n):
    for i in range(n):   #做3次(由外面給的值決定做幾次)
        time.sleep(0.5)  #暫停0.5秒
        print("Hello world "+ str(i))

def print_hi(n, interval):
    for i in range(n):
        time.sleep(interval) #暫停0.2秒(由外面給的值決定做幾次)
        print("Hi "+ str(i))

t1 = threading.Thread(target = print_hello)
t2 = threading.Thread(target = print_hellowrold, args = (3,))
t3 = threading.Thread(target = print_hi, args = (3, 0.2))

t1.start()
t2.start()
t3.start()
```
這程式只是在示範若是不用多執行緒而是一個一個做的話，總共需要0.5*3+0.5*3+0.2*3=3.6秒，但若使用多執行緒則只要1.5秒就好。
![螢幕擷取畫面 2022-03-03 151935](https://user-images.githubusercontent.com/99789312/156516036-062a2236-d2fa-4921-9065-70909dd5dc5c.jpg)
上面這張圖展示了確實3個函式是同時執行的，第一跟第二個函式還同時print。

PS 我用jupyter跑這程式時偶爾會有time.sleep休眠後喚不醒的問題(也就是打time.sleep(0.5)想休眠0.5秒，但程式卻一直休眠讓後面的程式都無法繼續執行)，若是有人知道是為什麼請告訴我，不然遇到這問題會很討厭。

**Comment by chia-kai-chang** [comment]
謝謝 @sean1325418 ，我來學一下~

看看可不可以利用這個技巧，同時擷取兩個溫度計的溫度讀值。

順便分享一個不錯的教學資源：https://blog.gtwang.org/programming/python-threading-multithreaded-programming-tutorial/

除了多執行緒之外，他還有介紹排隊(Queue)、鎖定(Lock)，介紹的相當完善，大家可以參考一下。

**Comment by pekoranoinu** [comment]
這個技巧我也要來學一學(≧ω≦)/

**Comment by boson13579** [comment]
之前為了跑演算法程式作業學了一下C++的多執行續的的用法，不過時間總是特別奇怪，後來查詢了一下發現原來c++的clock有坑，[蠻有用的教學](https://bbs.huaweicloud.com/blogs/157959?fbclid=IwAR1BJW9lTO6ZjhKyAYPdbe9GwqjC-bOznZ8fCRne-tiJBiuRssWFsxsi62g)，大家可以參考看看，是說之前沒學過python的，這次來學一下好了。
